File: R/example13_3.R
# Example 13.3 Berndt-Wood Model
# Berndt-Wood Factor Share Equations
# Basic Model: 3 Shares in relative price with restrictions
# Elasticities Interpretation

bwq<-read.table("http://web.pdx.edu/~crkl/ceR/data/bwq.txt",header=T,nrows=25)
bwp<-read.table("http://web.pdx.edu/~crkl/ceR/data/bwp.txt",header=T,nrows=25)
bwdata<-merge(bwp,bwq)
summary(bwdata)
rm(bwp,bwq)

attach(bwdata)
tc<-PK*K+PL*L+PE*E+PM*M
sk<-PK*K/tc
sl<-PL*L/tc
se<-PE*E/tc
sm<-PM*M/tc
# sk+sl+se+sm=1
# log prices are relative to PM (linear homogeneity)
lnpk<-log(PK/PM)
lnpl<-log(PL/PM)
lnpe<-log(PE/PM)
detach(bwdata)

# formula: share equations
fsk<-sk~lnpk+lnpl+lnpe
fsl<-sl~lnpk+lnpl+lnpe
fse<-se~lnpk+lnpl+lnpe
bwmodel<-list(SK=fsk,SL=fsl,SE=fse)

# setup linear constraints: symmetry
res1<-"SK_lnpl-SL_lnpk=0"
res2<-"SK_lnpe-SE_lnpk=0"
res3<-"SL_lnpe-SE_lnpl=0"

library(systemfit)
# Iterative SUR
out1<-systemfit(bwmodel,method="SUR",data=bwdata,maxiter=100,
                restrict.matrix=c(res1,res2,res3))
summary(out1)

# model interpretation based on elasticity
# elasticity of i (xk) w.r.t. j (pe)
# compute own price elasticities at the means
# compute cross price elasticities at the means
bij<-out1$coefficients["SK_lnpe"]
si<-as.matrix(fitted.values(out1)["SK"])
sj<-as.matrix(fitted.values(out1)["SE"])

elast_sij<-1+bij/(mean(si)*mean(sj))
elast_pij<-mean(sj)*elast_sij
names(elast_sij)<-"elast_sij"
names(elast_pij)<-"elast_pij"
c(elast_sij,elast_pij)

rm(list = ls())

File: python/example13_3.py
# Example 13.3 Berndt-Wood Model
# Berndt-Wood Factor Share Equations
# Basic Model: 3 Shares in relative price with restrictions
# Elasticities Interpretation

import numpy as np
import pandas as pd
from linearmodels import SUR

bwq = pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/bwq.txt',sep='\s+',nrows=25,index_col='YEAR')
bwp = pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/bwp.txt',sep='\s+',nrows=25,index_col='YEAR')
bwdata = pd.concat([bwp,bwq],axis=1)
print(bwdata.describe())

PK = bwdata.PK
K = bwdata.K
PL = bwdata.PL
L = bwdata.L
PE = bwdata.PE
E = bwdata.E
PM = bwdata.PM
M = bwdata.M
tc = PK * K + PL * L + PE * E + PM * M
sk = PK * K / tc
sl = PL * L / tc
se = PE * E / tc
# sm = PM * M / tc
# sk+sl+se+sm=1
# log prices are relative to PM (linear homogeneity)
lnpk = np.log(PK / PM)
lnpl = np.log(PL / PM)
lnpe = np.log(PE / PM)
# formula: share equations
fsk = 'sk~1+lnpk+lnpl+lnpe'
fsl = 'sl~1+lnpk+lnpl+lnpe'
fse = 'se~1+lnpk+lnpl+lnpe'

df1 = pd.DataFrame({'sk':sk,'sl':sl,'se':se,'lnpk':lnpk,'lnpl':lnpl,'lnpe':lnpe})
model1 = {'SK':fsk, 'SL':fsl, 'SE':fse}
model1_sur = SUR.from_formula(model1,data=df1)
# restriction matrix R: 3x12, Q: 3x1
cons_r1 = pd.DataFrame([[0,0,1,0,0,-1,0,0,0,0,0,0],
                        [0,0,0,1,0,0,0,0,0,-1,0,0],
                        [0,0,0,0,0,0,0,1,0,0,-1,0]])
cons_q1 = pd.Series([0,0,0])
model1_sur.add_constraints(cons_r1,cons_q1)
bwmodel1 = model1_sur.fit(iterate=True,cov_type='unadjusted')
print(bwmodel1.summary)
print('Asymptotic Variance-Covariance Matrix of Equations')
print(bwmodel1.resids.cov())

# model interpretation based on elasticity
# elasticity of i (xk) w.r.t. j (pe)
# compute own price elasticities at the means
# compute cross price elasticities at the means
bii = bwmodel1.params['SK_lnpk']
bij = bwmodel1.params['SK_lnpe']
si = bwmodel1.fitted_values.SK.mean()
sj = bwmodel1.fitted_values.SE.mean()

elast_sij = 1 + bij / si / sj
elast_pij = sj * elast_sij
elast_sii = 1 + (bii - 1) / si / si
print(elast_sij)
print(elast_pij)
